enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,232 @@
+
+### Goals
+
+- Prepare openshift components to accomodate the out-of-tree plugins and remove support for the in-tree implementation.
+- Describe an approach to select an `external` cloud provider installation for the cloud.
+- Provide the means to allow generic `cloud-controller` component registration via the provider specific upstream implemetation.
+
+### Non-Goals
+
+- Force immediate exclusion of in-tree support for currently used providers, as their out-of-tree counterparts are added.
+- Add support for CSI and other out-of-tree implementations of cluster storage interfaces.
+- Add support for running multiple cloud configurations in a single cluster.
+
+## Proposal
+
+Add an out-of-tree support based on current IPI with minor adjustments, and make a seamless transition for all currently supported providers.
+
+This change will be gradually executed on all suported cloud providers: AWS, Azure, GCP, vSphere and IBM.
+
+### User Stories
+
+As a cloud developer, I’d like to improve support for cloud features for openshift
+
+#### Story 1:
+
+We’d like to deploy control-plane node on a spot instance, to allowing customers to get additional cost reduction on their clusters. In order to introduce and test this feature, we have to implement this in the kubernetes repository, first, then wait for the next release to get a “free” rebase on the feature. We are also bound to the kubernetes release cycles, making it hard to achieve the goal in case the core Kubernetes is already going through the feature freeze. Starting downstream is not an option, as the implementation could be rejected upstream or gain less attention then needs to quickly get it into the release, diverging our fork from the source.
+
+#### Story 2:
+
+As an openshift developer I’m assigned a release blocker BZ regarding upstream implementation. Proving the value for merging this and making the bug a release blocker upstream requires careful communication with upstream maintainers, and explaining the value to the people whom are not necessarily involved into internal details, so their opinion could differ and delay merging an important fix.
+
+#### Story 3:
+
+We desire to extend our team responsibilities on upstream cloud providers in the future and gain weight in promoting features into upstream. Getting such for kubernetes repository is harder for us and maintainers, as this would mean giving weight for approving and merging features for the parts of the kubernetes project, not necessarily laying under our responsibilities and scope of knowledge.
```

Comment:
This is not correct. Even already each sig/community member can work and does work only on a specific area of code and can easily be reviewer, approver and owner of that code.